modules/packages/OrderedSet.chpl

```diff
@@ -0,0 +1,484 @@
+/*
+ * Copyright 2004-2020 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This module contains the implementation of the orderedSet type.
+
+  An orderedSet is a collection of unique ordered elements.
+
+  All references to orderedSet elements are invalidated when the orderedSet is 
+  cleared or deinitialized.
+
+  orderedSet are not parallel safe by default, but can be made parallel safe by 
+  setting the param formal `parSafe` to true in any orderedSet constructor.
+
+  When constructed from another orderedSet, 
+  the new orderedSet will inherit the parallel safety mode 
+  of its originating orderedSet.
+
+  orderedSet is designed with the ability to use an enum to choose from different 
+  implementations under the hood. For now, only treap is supported.
+*/
+module OrderedSet {
+  include module Treap;
+  private use Treap;
+  private use Reflection;
+  private use IO;
+  public use Sort only defaultComparator;
+
+
+  //TODO: Maybe add skipList in the future
+  /* Implementations supported */
+  enum setImpl {treap};
+
+  pragma "no doc"
+  proc getTypeFromEnumVal(param val, type eltType, param parSafe) type {
```

Comment:
In these functions that return something if `val == setImpl.treap`, and don't return anything otherwise, I'd prefer there be a `compilerError()` in the returning-nothing case.  Or at least specify the formal argument as `param val: setImpl`.

Reply:
https://github.com/chapel-lang/chapel/pull/16124/commits/4aa87b92c888358a4a20731a8ce2674668d34317 specified here